### DISCUSSION TITLE
How do I dual boot with custom main.py?

### DISCUSSION THREAD

**Discussion by fatherchaos** [pybricks/support#478] [Q&A]
Hi,
I'm trying to flash a dual boot firmware to the PrimeHub with a modified main.py. My goal is to load a custom main.py to use when the unit boots to pybricks and charge the unit in the regular firmware. This is intended for first lego league.

I can use the website install.pybricks.com to install the dual boot firmware, but there is no option to select my custom firmware.

When I use "py pipx run pybricksdev -flash primehub.zip" the primehub must be in dfu and does not install the dual boot firmware. I found that oybricksdev had a -dual-boot argument in January, but the command does not appear to work now.

Can someone guide me through how to install a modified dual boot firmware with custom main.py?

**Comment by laurensvalk** [comment]
> Can someone guide me through how to install a modified dual boot firmware with custom main.py?

This is not currently supported. Since the Prime Hub has 30Mb of storage space, our plan is to enable that instead of having to flash something every time.

Though I can see that this is not too helpful in the short term.

In case you have more than one set, is it an option to use the second Prime Hub to charge the battery in, and then swap them?

**Reply by fatherchaos** [reply]
Charging on another unit won't be possible unless I can maintain the dual boot feature. Both bots are used by the kids and need to be identicle.

Currently I have been successful with modifying main.py in the primehub-firmware.zip. When I flash the program I loose the dual boot option but get the non volatile program storage I need. I know this solution is non ideal but it meets our short term needs.

I don't mind a convoluted flash option for now. I assume my best options would be a choose file feature from install.pybricks or to downgrade to the pybricksdev that had the dual-boot option from January. Perhaps I can host our own version of the website and point towards a custom main.py file to zip?

If you can point me in the right direction I'll do the leg work.

**Reply by dlech** [reply]
Maybe we could add a *Browse...` button to https://install.pybricks.com/ like we have on https://dfu.pybricks.com?

I'm not sure if the older version of pybricksdev would still do dual-boot the same as it is now, but if it does, that would probably be the simplest option.

Otherwise, you could build from source ([instructions](https://github.com/pybricks/pybricks-micropython/blob/master/CONTRIBUTING.md)). When you get to the *Building the code* part, you will need to copy your `main.py` to `bricks/primehub/main.py`, then with the hub plugged in with USB in DFU mode (start from powered off, press and hold the Bluetooth button, plug in USB, wait for light to turn purple, release Blueooth button) run `make -C bricks/primehub deploy-dfu-dual-boot BASE_FW=official.bin` where `official.bin` is the official LEGO firmware that you backed up or got from https://github.com/gpdaniels/spike-prime/tree/master/firmware.

**Reply by fatherchaos** [reply]
I think a browse button from install.pybricks.com would be the best option for now. Is the source for this available? I can try to copy the functionality from dfu.pybricks.com if someone can point me to the right repo.

**Reply by laurensvalk** [reply]
I don’t think any of our dual boot methods currently include a main.py script.

We never bothered to make that work, because we would eventually support file storage. But maybe it is worth the effort in the mean time.

**Reply by fatherchaos** [reply]
If you can provide some initial guidance I'm willing to put in the time to get something basic going. I did see in the script2.js from https://install.pybricks.com/ that there is an uploadfile method. Would it be possible to use uploadfile() and send main.py and have the firmware seek only that file on boot?

**Reply by dlech** [reply]
The `main.py` file has to be compiled to a `.mpy` file and appended to the Pybricks firmware blob. [This is how Pybricks Code does it.](https://github.com/pybricks/pybricks-code/blob/master/src/firmware/sagas.ts#L178-L262)

**Reply by fatherchaos** [reply]
So if I have this right the workflow would be:

1. pybricksdev compile main.py to generate .mpy bytecode
-- In our case the project will need to be flattened to one file, as there are other .py files in use to help the kids keep everything clean. I saw a discussion on inlining multiple files somewhere, so I'll look for a similar utility.
2. Run loadfirmware to combine the .mpy with firmware-dual-boot-base.bin, subject to 30meg file size limit. 
--I don't see us going over this limit as the code will be rather small in size. Most of our code is just the same commands performed in different sequences.
3. Select firmware file somehow on install.pybricks.com
4. Install...

**Reply by dlech** [reply]
A modified install.pybricks.com should handle most of it for you. So the workflow would just be update the `main.py` in the `firmware.zip` file then flash the firmware to the hub.

**Reply by fatherchaos** [reply]
I'm still learning where all the code repos are... can you point me towards where install.pybricks & dfu.pybricks is located? I assume I can copy most of the file select method from dfu.pybricks.

**Reply by laurensvalk** [reply]
There may be a bit more to this before we go there. I don't think the dual-boot firmware can currently find the `.mpy` file even if it was included, because it doesn't know where to look.

So if we're going to do this, the [firmware](https://github.com/pybricks/pybricks-micropython) is probably the first thing to start with. Once it works with the [`make`](https://github.com/pybricks/support/discussions/478#discussioncomment-1324896) command, it is possible to start thinking about a `Pybricksdev` update or the web installer.

Or, we focus our efforts on the most minimal implementation of file storage to save a script as it is downloaded to the hub :)

**Reply by fatherchaos** [reply]
The file storage implementation seems the best path as that is the eventual goal. Is it currently possible for the firmware to see files stored in _pybricks/ ? I assume we can have rudimentary support to look for main.py or main.mpy. In our current scenario having only one file will work.

For now I assume the operation could be like this:
Flash dual boot firmware from install.pybricks
utility to upload main.mpy ( uploadFile('main.mpy') )
utility to remove _pybricks/main.mpy (uos.remove('_pybricks/main.mpy)
On boot, check for main.mpy and load. If absent run normally to support beta.pybricks.com

**Reply by fatherchaos** [reply]
Hi, I just want to make sure I'm on track with this.

I'm hoping to have a short term solution available for the kids to help them test the primehub using pybricks. What would be the simplest option to have the firmware find a main.py or main.mpy? If I can compile/upload one python file that might be the fastest for the students. Right now I envision the easiest path would be to use the os upload function and just overwrite the existing main.py or main.mpy file stored in _pybricks/.

I'd like to help with the rudimentary file storage for the firmware, but do not know where to start.

In the meantime I have made a local copy of install.pybricks.com to make a hacked together solution to compile the firmware with pybricksdev -compile and use that file (with updated hash) on my local copy of install.

**Reply by laurensvalk** [reply]
I don't think we'll be able to do this in the short term, I'm afraid.

Is it an option to just use dual boot with [Pybricks Code](code.pybricks.com) most of the time, and only use `pybricksdev` to occasionally test the single-boot-with-script approach, just to make sure things are working for competition day?

> Right now I envision the easiest path would be to use the os upload function and just overwrite the existing main.py or main.mpy file stored in _pybricks/.

There's only the installation script in that folder, there is no `main` script.

**Reply by fatherchaos** [reply]
Right now we are using the dual boot from pybricks code (beta). This works well enough for now for the little bits they are making. We plan to use pybricksdev to flash with the full program as you mentioned.

My thought was to upload the main.py using the same method that the install script from install.pybricks does, snippet below for reference:
```
case 1:
                    _a.sent();
                    return [4 /*yield*/, uploadFile('install_pybricks.py')];
```
For this case I'd make a simple button to perform the following:
```
case 1:
                    _a.sent();
                    return [4 /*yield*/, uploadFile('main.py')];
```

which would create the file _pybricks/main.py

If the firmware can see that file then I can stick with a single upload for main.py or a compiled main.mpy

**Reply by laurensvalk** [reply]
> If the firmware can see that file then I can stick with a single upload for main.py or a compiled main.mpy

To reiterate, there is currently nothing in the firmware that can read files from file storage.

But there's some good news as well. We've been making some initial progress on saving the program as it gets downloaded to the hub.

If all goes well, that means students can just use Pybricks Code and dual boot without further modifications.

**Reply by fatherchaos** [reply]
That sounds great. Is there any way I can help?
